### PR TITLE
QuLDE solvers

### PR DESCRIPTION
**Pull request by dgan181**
Included both HHL (Multi-Linear step) and Non-HHL method.

### CODE REVIEW

File: src/QuDiffAlgs.jl [hunk 1]
```diff
@@ -0,0 +1,47 @@
+export QuLDE, LDEMSAlgHHL
+export QuEuler, QuLeapfrog, QuAB2, QuAB3, QuAB4
+
+abstract type QuODEAlgorithm <: DiffEqBase.AbstractODEAlgorithm end
+#abstract type QuODEProblem{uType,tType,isinplace} <: DiffEqBase.AbstractODEProblem{uType,tType,isinplace} end
+abstract type LDEMSAlgHHL <: QuODEAlgorithm end
+
+
+#Linear differential equation solvers (non-HHL)
+struct QuLDE <: QuODEAlgorithm end
```
Comment (Roger-luo) [inline src/QuDiffAlgs.jl]: Can we have docstring here? at a reference to the paper.

File: src/QuDiffHHL.jl [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+"""
+
+"""
+    LDEMSAlgHHL
+    * step - step for multistep method
+    * α - coefficients for xₙ
+    * β - coefficent for xₙ'
+"""
+
+function bval(alg::LDEMSAlgHHL,t,h,g::Function)
```
Comment (Roger-luo) [inline src/QuDiffHHL.jl]: Is there any reason why the function arg is the last arg? if it is the first arg, then we could use `do` block later.

**Comment by Roger-luo** [comment]
So before we move everything into this, I'm wondering if we could move HHL (that's the only dep you need from AlgorithmZoo I believe?) to this repo first. And you probably need to polish the HHL implementation a little bit, cause it might be messy at the moment.

File: src/QuDiffHHL.jl (hunk 2 above)
Comment (dgan181) [reply src/QuDiffHHL.jl]: No, no reason. I don't think I can use `do` here. `g` is a function we get from the user (It is input vector 'b' as a function of time in time, in this case).

**Comment by dgan181** [comment]
Sure. Where do you suggest I start in the HHL implementation ?

File: src/QuDiffHHL.jl (hunk 2 above)
Comment (Roger-luo) [reply src/QuDiffHHL.jl]: Then do block is even more necessary, since it is a syntax sugar for user to input large functions.

Comment (dgan181) [reply src/QuDiffHHL.jl]: Uh. I am not sure how this is to be done. Can you guide me a little?

Comment (Roger-luo) [reply src/QuDiffHHL.jl]: Simply make the function as your first argument. Maybe reading the doc of do block would help?

Comment (dgan181) [reply src/QuDiffHHL.jl]: I realise how it is to be done. I've made the change. Please let me know if you think it's fine.

**Comment by Roger-luo** [comment]
Maybe we could start with copy-pasting HHL here first? I think we don't want an unstable dep at least.

**Comment by Roger-luo** [comment]
I'll enable CI for the repo a bit later.

**Comment by dgan181** [comment]
HHL alone is not the only dep. I'll need to move the QFT and PE code as well (maybe more) . Perhaps we shouldn't move HHL ?

**Comment by Roger-luo** [comment]
The thing is the other part of QuAlgorithmZoo is not very stable, e.g we add new algorithms, change the API due to new algorithms etc. 

That repo is mainly for demonstration and examples.

If we release this package after gsoc it would make this package unstable as well…we could do this later. I'll help on this.

**Comment by dgan181** [comment]
I see. I'll start copying all the files I need. If I have any issues with the dependancies I'll let you know.
Have you had chance to go through my Qu LDE (non-HHL) code?

**Comment by Roger-luo** [comment]
It looks good to me. But I really don't have chance to review the code in detail this week. I could merge it first since the test seems to work. We could revise it anyway if we hit issues.

**Comment by Roger-luo** [comment]
And I think you missed a type parameter in general matrix block? The number of qubits?

File: src/QuDiffHHL.jl (hunk 2 above)
Comment (Roger-luo) [reply src/QuDiffHHL.jl]: Yes, that's basically how you do it. Yes I think this is good now.

**Comment by Roger-luo** [comment]
The CI is broken because you are `using QuAlgorithmZoo` but it is deleted in Project.toml now. Fix the CI test and I think we could merge it then.

File: src/QuLDE.jl [hunk 3]
```diff
@@ -0,0 +1,89 @@
+
+struct QuLDEParam{W,L,Q}
+    k::Int
+    t::L
+    C_tilda::W
+    D_tilda::W
+    N::W
+    V::GeneralMatrixBlock{1}
```
Comment (Roger-luo) [inline src/QuLDE.jl]: This is not correct here, the type parameter of GeneralMatrixBlock is

```julia
"""
    GeneralMatrixBlock{M, N, MT} <: PrimitiveBlock{N}
General matrix gate wraps a matrix operator to quantum gates. This is the most
general form of a quantum gate. `M` is the hilbert dimension (first dimension),
`N` is the hilbert dimension (second dimension) of current quantum state. For
most quantum gates, we have ``M = N``.
"""
struct GeneralMatrixBlock{M, N, T, MT <: AbstractMatrix{T}} <: PrimitiveBlock{N}
```

If you don't want to write these type parameters just write it in this way:

```
struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixBlock{Q}}
    k::Int
    t::L
    C_tilda::W
    D_tilda::W
    N::W
    V::VType
   VS1::SType
    VS2::SType
    WS1::SType
    WS2::SType
end
```

File: src/QuDiffHHL.jl [hunk 4]
```diff
@@ -0,0 +1,102 @@
+export array_qudiff, prepare_init_state, bval, aval
+import QuAlgorithmZoo.hhlsolve
```
Comment (Roger-luo) [inline src/QuDiffHHL.jl]: Have you moved HHL to this repo or do we still need this dep? if we do, you need to add the QuAlgorithmZoo back to make CI pass first.

Comment (dgan181) [reply src/QuDiffHHL.jl]: I added QuAlgorithmZoo and included the dep inside `Project.toml` .

File: src/QuLDE.jl (hunk 3 above)
Comment (dgan181) [reply src/QuLDE.jl]: I used the second thing you suggested.

**Comment by dgan181** [comment]
@Roger-luo I made this repo independent of QuAlgorithmZoo. I think we can merge this now.

File: src/QFT.jl [hunk 5]
```diff
@@ -0,0 +1,63 @@
+@static if VERSION >= v"0.7-"
```
Comment (Roger-luo) [inline src/QFT.jl]: BTW, I think we could delete this, since nobody will use v0.7 anymore.

**Comment by Roger-luo** [comment]
Nice, I think we will need a short blog post next week, maybe could have some demo on solving a simple ODE function, like a parabola? Or anything you have in mind?

**Comment by dgan181** [comment]
Sure. I have started working on the blog post. This method can only be used for exponential equation (diff eps of the form : x' = Mx + b). Till now I have written code for when M is unitary. I need to add the case for when M in non-unitary. I shall do hopefully finish that soon. I was thinking I could write 2 posts one explaining the unitary case and the other, the non unitary case.

**Comment by Roger-luo** [comment]
Yeah, I'm wondering if you are interested in doing non-linear form, we could linearize a nonlinear ODE so it can be solved approximately with this algorithm, so we could make this work for more practical problems.

But let's finish the demo first, and let me know if you need me to polish the blog post.

**Comment by dgan181** [comment]
Oh! Yes. So, I'll include that in the blog then. I'll ask you to review it as soon as my first draft is complete. Thanks.

Btw, (just to remind you) the next algorithm on my proposal is a non-linear diff eq solver (quadratic), but it uses the forward Euler method.

**Comment by Roger-luo** [comment]
Yes, thanks for reminding, I was meaning more general non-linear form that can be linearized to make use these algorithms. This is because I just recall I had an idea using quantum circuit to solve a Landau damping problem a few years ago, it should be straight forward with these solvers (I guess), so we might be able to make a paper later as well.

**Comment by dgan181** [comment]
This is going to be exciting.

**Pull request merged**